### ISSUE TITLE
Linux router issues when `ip route add` fails

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#2730]
In a few logs I've debugged lately (both 1.14.0), I see e.g.:

dd9683bf6d3a69853d87dbc9356d54ac5e1de41ccadd8fc15bb6fbf82607b6c2
```
2021-08-29 10:50:06.501602413 +0300 +0300: health("router"): error: running "ip route add fe80::/64 dev tailscale0 table 52" failed: exit status 2
RTNETLINK answers: File exists
2021-08-29 10:50:06.501625713 +0300 +0300: control: controlclient: restarting map request for "router" health change to new state: running "ip route add fe80::/64 dev tailscale0 table 52" failed: exit status 2
RTNETLINK answers: File exists
```

2ea3906c9327f6c70fd67a802ccebbd986e5b23c7585c88a7d7f63499508de69
```
2021-08-27 16:43:21.284122602 -0600 -0600: router: localRoute add failed: running "ip route add throw fe80::/64 table 52" failed: exit status 2
RTNETLINK answers: File exists
```

Investigate both, figuring out how to repro.

Maybe when IPv6 is force disabled?

Maybe on unclean shutdown? (does anything clean up those throw rules on start-up?)

/cc @DentonGentry @danderson 

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_3imz5)

**Comment by danderson** [comment]
Very confused, because I have no memory of adding link-local routes anywhere in the router implementation. The kernel does that automatically, so I can't think why that would ever be necessary or useful.

**Comment by danderson** [comment]
Based on that log line, allegedly we're telling wgengine/router that we want to add a route for fe80::/64. I can't find anywhere in our codebase where we do this. My first instinct is that this is some kind of malfunction involving attempts to restore stuff that we think got deleted by the kernel incorrectly, but I haven't traced through our spaghetti yet to find where that might be.

**Comment by bradfitz** [comment]
> My first instinct is that this is some kind of malfunction involving attempts to restore stuff that we think got deleted by the kernel incorrectly, but I haven't traced through our spaghetti yet to find where that might be.

A good theory, but its log print never shows up in those logs above:

```go
   r.logf("somebody (likely systemd-networkd) deleted ip rules; restoring Tailscale's")
````

So it doesn't seem to be `func (r *linuxRouter) onIPRuleDeleted` at fault.

**Comment by DentonGentry** [comment]
One of the nodes logging
`2021-08-27 16:43:21.284122602 -0600 -0600: router: localRoute add failed: running "ip route add throw fe80::/64 table 52" failed: exit status 2`

has IPv6 completely disabled.

If I'm reading https://github.com/tailscale/tailscale/blob/a59b389a6a61fd27cd41acd15effae9e2e751697/wgengine/router/router_linux.go#L279 correctly, having the `ip route add throw` fail will cause linuxRouter.Set to return early.

**Comment by bradfitz** [comment]
@DentonGentry, it can't "return early" because there's exactly 1 return statement in that whole method: on the last line.

It can return with an error for sure (as we see it does), but at least the other 4 methods after it still run, accumulating their errors.

**Comment by DentonGentry** [comment]
Calls to router.Set() that I can find, and the impact if it returns failure:

https://github.com/tailscale/tailscale/blob/640134421eb922e8b85c3686f9966d7001a02879/wgengine/userspace.go#L399
would not start linkmonitor.

https://github.com/tailscale/tailscale/blob/640134421eb922e8b85c3686f9966d7001a02879/wgengine/userspace.go#L878
would not set DNS config.

**Comment by bradfitz** [comment]
Ah, the _callers_ return early.

**Comment by bradfitz** [comment]
@darshinimashar, which Ubuntu 20.04 variant are you using? Desktop, server, cloud-init?

**Comment by pcrockett** [comment]
I started experiencing this today, on two separate installations of Fedora Silverblue. Except I'm noticing the error message when running `tailscale status`.

Two additional observations:

1. This only happens when I'm using an exit node, a VPS running on Linode.
2. When I use `--accept-routes`, the IP address in the error message changes to the IPv4 subnet for which I've configured a subnet router.

**Comment by DentonGentry** [comment]
Tailscale 1.18 no longer runs the `ip` command, using netlink messages instead.

**Comment by justinengland** [comment]
when does 1.18 release?

**Comment by DentonGentry** [comment]
Soon. We have one blocking issue left, a connectivity glitch. We'll evaluate the fix for that tomorrow, and if it works then we could release 1.18 on Monday November 15.

**Comment by justinengland** [comment]
any update on this release?

**Comment by DentonGentry** [comment]
1.18 released two weeks ago, with a 1.18.1 release fixing one issue with the RTNETLINK work discussed here.

https://github.com/tailscale/tailscale/releases/tag/v1.18.0
https://github.com/tailscale/tailscale/releases/tag/v1.18.1

**Comment by justinengland** [comment]
Thank you very much!

**Comment by DentonGentry** [comment]
Fixed in 1.18.1.
